### ISSUE TITLE
Error in signs of scaling when decomposing matrix 

### ISSUE THREAD

**Issue by davidyxu** [paperjs/paper.js#1391]
# Description/Steps to reproduce
1. Scale a paper.item by (-1, 1)

# Link to reproduction test-case
http://sketch.paperjs.org/#S/jZFNasMwEIWvMszKBickBm9Uuio9QZaRF7IyTkQmGiMZShty98qxQlNCf6SN5unNmw/pjN6cCBVujjTaA1ZoZTfVVnwcwbpgmeAZPL3B5mAGWr4yuyFScdYe0uod84uwBAUaA+00VvPFIM6PCrbNqoJm1WY1ug/KYt202l9K7ad9P65O8+bT0rJ4Ksqnm0WSxrIvNEZr2Pk9dO9QLNYVrEuNk++7q6NeAqkEdUvMfVfrl0I54yHA9COFH/p/Q0ppi38i1Y9MdYaaU/6EuktIH9gFMsfr60dU2/byCQ==
# Expected result
After scaling by (-1, 1) `circle.scaling` should be (-1, 1), not (1, -1).

# Additional information
macOS 10.11.5
Version 61.0.3163.100 (Official Build) (64-bit)

**Comment by lehni** [comment]
You're right... Here a [version of the sketch](http://sketch.paperjs.org/#S/rZLBS8MwFMb/lWcua6Erm9hLxJN4FAQPHpYd0vS1DY15JcnUOfq/265hMpiIYG75fV8+3hfegVn5ioyz5w6DalnGFFXTXZH1AZR2yiDcgcV3eJKhzR+M0b3H5CAsjKcnbQOHTbHKoFhts5l6/YkRXhcnGBx1eE+GHAfBHFaCnUkvugoth5tzv+w5LBztbLWIgux7s3+UwekPDrU0HqOgDHmsIhN2SIUV9k06CNQ0xxrB7XCCuobkaqYpxCpTYzKYG2oSwbySRtsGyj0ky3UG61Sw9PaSs8SaHPKxTPyuPL492b8pxqyLQbIO6H7IGQDHTr+POqYv/2vUOevvowo7rlHpUHbH7fCMb7bDFw==) for easy toggling and displaying the difference.

**Comment by sasensi** [comment]
After investigating on this one, I found that this is related to how we decompose matrix.
I'm not sure that this is a bug because if you also check `circle.rotation`, you'll see that it is equal to `180°`. And `scale(-1,1)` is equivalent to `scale(1,-1) + rotate(180)`. So transformation is correct (what we can confirm by looking at the drawing or by checking the matrix).
Following the links in `matrix.decompose()` comments:
https://github.com/paperjs/paper.js/blob/655a4dabd005f222f57b93904e1199d53db73b2b/src/basic/Matrix.js#L681-L684
I found that we are actually using a decomposition algorithm known as [QR decomposition](https://en.wikipedia.org/wiki/QR_decomposition) which for a given `matrix(-1, 0, 0, 1, 0, 0)` outputs `scale(1,-1) + rotate(180)`.
This [page](http://frederic-wang.fr/decomposition-of-2d-transform-matrices.html) pretty well explain how it works and allow trying different inputs/outputs. Actually, it also presents an alternative algorithm called [LU decomposition](https://en.wikipedia.org/wiki/LU_decomposition) which for a given `matrix(-1, 0, 0, 1, 0, 0)` outputs `scale(-1,1)` as we expect here.
So it seems that switching to LU algorithm would solve this case but it may also create similar surprising outputs for other transformations...

@lehni and @sapics, what do you think ?

**Comment by sapics** [comment]
Welcome back @sasensi! Sorry for late response.
I guess that it would be possible to fix by adding this.
```js
if (!b && !c) {
    scale = [a, d];
    rotate = 0;
    skew = [0, 0];
}
```

**Comment by sasensi** [comment]
@sapics, thanks !
Unfortunately, I think this one is unsolvable because even with your fix, we would still have surprising results when using `rotate()` too.
Try [this](http://sketch.paperjs.org/#V/0.11.8/S/xVPBSgMxEP2VMZfuwnZpxV4insSjIPTgodtDmp3dho2ZJUnVWvrvJt3ttlQED6KBJOTNvJl5D7JjRrwg42zeoJdrljFJZXxLMs6DVFZqhDsw+AZPwq/zB61V6zDZFbYwEFZLyngOi9kkg9lkmR1xpz6wh69nZ7C31OA9abIcCmaxLNhF8FmVfs3h5pIjWg4jSxtTjoaQaFu9fRTeqncOldAOh5DU5LDs0Qju09t4xf0qLHiq64M2bzc4RDrFuSUvPCbTyYmjKkiuOlIKg/zoEwWCpjopmJNCK1PDagvJeJrBNC1YV+Fr7gorsgg9hQcbervzHso9zYMwUydpFoWHiRQZPhryjtBZh1MB7Af4pruoPNrfa74HDC7/xJYw1PhfbekG+CNb4hE+1cqiaA4/xTG+WO4/AQ==) with your fix applied.

**Comment by sapics** [comment]
I agree ;)
As you noticed, we can get multiple affine transformations for representing one matrix, so both result `scale(-1,1)` and `scale(1,-1)` are right. It looks difficult to avoid surprise :)
